AppCenterIdentity/AppCenterIdentity/Internals/Storage/MSKeychainAuthTokenStorage.m

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#import "MSKeychainAuthTokenStorage.h"
+#import "MSAuthTokenInfo.h"
+#import "MSIdentityConstants.h"
+#import "MSIdentityPrivate.h"
+#import "MSKeychainUtil.h"
+#import "MSLogger.h"
+#import "MSUtility.h"
+
+@implementation MSKeychainAuthTokenStorage
+
+- (nullable NSString *)retrieveAuthToken {
+  NSString *authToken = [MSKeychainUtil stringForKey:kMSIdentityAuthTokenKey];
+  if (authToken) {
+    MSLogDebug([MSIdentity logTag], @"Retrieved auth token from keychain.");
+  } else {
+    MSLogWarning([MSIdentity logTag], @"Failed to retrieve auth token from keychain or none was found.");
+  }
+  return authToken;
+}
+
+- (nullable NSString *)retrieveAccountId {
+  return [MS_USER_DEFAULTS objectForKey:kMSIdentityMSALAccountHomeAccountKey];
+}
+
+- (MSAuthTokenInfo *)oldestAuthToken {
+
+  // Read token array from storage.
+  NSMutableArray<MSAuthTokenInfo *> *tokenArray = [MSKeychainUtil arrayForKey:kMSIdentityAuthTokenArrayKey];
+  if (tokenArray.count == 0) {
+    return nil;
+  }
+
+  MSAuthTokenInfo *authTokenInfo = tokenArray.firstObject;
+  NSDate *nextChangeTime = tokenArray.count > 1 ? tokenArray[1].startTime : nil;
+  if ([authTokenInfo.endTime laterDate:nextChangeTime]) {
+    return [[MSAuthTokenInfo alloc] initWithAuthToken:authTokenInfo.authToken
+                                         andStartTime:authTokenInfo.startTime
+                                           andEndTime:nextChangeTime];
+  }
+  return authTokenInfo;
+}
+
+- (void)saveAuthToken:(nullable NSString *)authToken withAccountId:(nullable NSString *)accountId expiresOn:(nullable NSDate *)expiresOn {
+  @synchronized(self) {
+
+    // Read token array from storage.
+    NSMutableArray<MSAuthTokenInfo *> *tokenArray = [MSKeychainUtil arrayForKey:kMSIdentityAuthTokenArrayKey];
+    if (tokenArray.count == 0) {
+
+      // Add nil token if the entire story is empty.
+      [tokenArray addObject:[MSAuthTokenInfo new]];
```

Comment:
Can you explain this? Will there just always be nil token at the first place?

Reply:
Yes, right. We're adding a `nil` token if it's a first element in the history (only occurs once during the first initialization of history). This will be used later for querying data.